python_modules/libraries/dagster-embedded-elt/dagster_embedded_elt/dlt/resource.py

```diff
@@ -99,26 +99,45 @@ def extract_resource_metadata(
     def run(
         self,
         context: Union[OpExecutionContext, AssetExecutionContext],
+        dlt_source: Optional[DltSource] = None,
+        dlt_pipeline: Optional[Pipeline] = None,
+        dagster_dlt_translator: Optional[DagsterDltTranslator] = None,
         **kwargs,
-    ) -> Iterator[Union[AssetMaterialization, MaterializeResult]]:
+    ) -> Generator[Union[MaterializeResult, AssetMaterialization], None, None]:
         """Runs the dlt pipeline with subset support.
 
         Args:
             context (Union[OpExecutionContext, AssetExecutionContext]): Asset or op execution context
+            dlt_source (Optional[DltSource]): optional dlt source if resource is used from an `@op`
+            dlt_pipeline (Optional[Pipeline]): optional dlt pipeline if resource is used from an `@op`
+            dagster_dlt_translator (Optional[DagsterDltTranslator]): optional dlt translator if resource is used from an `@op`
             **kwargs (dict[str, Any]): Keyword args passed to pipeline `run` method
 
         Returns:
-            Iterator[Union[AssetMaterialization, MaterializeResult]]: A generator of AssetMaterialization or MaterializeResult
+            Generator[Union[MaterializeResult, AssetMaterialization], None, None]: A generator of MaterializeResult or AssetMaterialization
 
         """
-        metadata_by_key = context.assets_def.metadata_by_key
-        first_asset_metadata = next(iter(metadata_by_key.values()))
+        # This resource can be used in both `asset` and `op` definitions. In the context of an asset
+        # execution, we retrieve the dlt source, pipeline, and translator from the asset metadata.
+        # This allows us to provide the parameter _only_ in the `dlt_assets` decorator.
+        if isinstance(context, AssetExecutionContext):
+            metadata_by_key = context.assets_def.metadata_by_key
+            first_asset_metadata = next(iter(metadata_by_key.values()))
+
+            dlt_source = check.inst(first_asset_metadata.get(META_KEY_SOURCE), DltSource)
+            dlt_pipeline = check.inst(first_asset_metadata.get(META_KEY_PIPELINE), Pipeline)
+            dagster_dlt_translator = check.inst(
+                first_asset_metadata.get(META_KEY_TRANSLATOR), DagsterDltTranslator
+            )
+
+        else:
+            if dlt_source is None or dlt_pipeline is None:
+                raise ValueError(
+                    "dlt_source, and dlt_pipeline parameters must be provided in an `op` context"
+                )
```

Comment:
Use check.invariant

Reply:
Idk why I struggle so much with the `check` module and type checking, but using `check.invariant` I get warnings downstream that the variable can possibly be none when using:

```
check.invariant(not (dlt_source and dlt_pipeline))
```

Am I making a silly boolean logic mistake? 😅 

```
>>> for (a, b) in [(None, None), (None, 1), (1, None), (1, 1)]:
...     try:
...         _ = check.invariant(not (a and b))
...         _c = True
...     except:
...         _c = False
...     print(a is None or b is None, _c)
...
True True
True True
True True
False False
```

Reply:
I updated it to the following which works as expected.
```python
dlt_source = check.not_none(
    dlt_source, "dlt_source is a required parameter in an op context"
)
dlt_pipeline = check.not_none(
    dlt_pipeline, "dlt_pipeline is a required parameter in an op context"
)
```